edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,725 @@
+```rust
+    Uuid(Uuid), // from https://docs.rs/uuid/latest/uuid/
+    Bytes(Bytes), // from https://docs.rs/bytes/latest/bytes/
+```
+
+Variants holding other Values:
+
+```rust
+    Set(Vec<Value>),
+    Object { shape: ObjectShape, fields: Vec<Option<Value>> },
+    SparseObject(SparseObject),
+    Tuple(Vec<Value>),
+    NamedTuple { shape: NamedTupleShape, fields: Vec<Value> },
+    Array(Vec<Value>),
+    Range(Range<Box<Value>>),
+```
+
+## Alternatives to working with Value
+
+There are a lot of alternatives to Value when dealing with the output from EdgeDB on the Rust side.
+
+### Using json
+
+Using json is pretty comfortable for Rust user thanks to serde and serde_json. EdgeDB can cast any type to json with `<json>` so just sticking that in front of a query is enough to return the same object as json:
+
+```rust
+let query = "select <json>(
+    insert Account {
+    username := <str>$0
+    }) {
+    username, 
+    id
+    };";
+
+// We know there will only be one result so use query_single_json;
+// otherwise it will return a map of json
+let json_res = client
+    .query_single_json(query, &("SomeUserName",))
+    .await?
+    .unwrap();
+```
+
+You can turn this into a serde_json::Value and access using square brackets:
+
+```rust
+let as_value: serde_json::Value = serde_json::from_str(&json_res)?;
+println!(
+    "Username is {},\nId is {}.\n",
+    as_value["username"], as_value["id"]
+);
+```
+
+But deserializing into a Rust type is much more common (and rigorous). If you have an Account struct that implements Deserialize, you can use serde_json to deserialize the result into an Account.
+
+```rust
+#[derive(Debug, Deserialize)]
+pub struct Account {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let as_account: Account = serde_json::from_str(&json_res)?;
+```
+
+### Using the Queryable macro
+
+The edgedb-derive crate has a built-in Queryable macro that lets us just query without having to cast to json. Same query as before:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct QueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+let query = "select (
+    insert Account {
+    username := <str>$0
+    }) {
+    username, 
+    id
+    };";
+let as_queryable_account: QueryableAccount = client
+    .query_required_single(query, &("SomeUserName",))
+    .await?;
+```
+
+Note: field order (in other words, the "shape" of the query) matters when using the Queryable macro. In the example before, a query is done in the order `id, username` instead of `username, id` as defined in the struct:
+
+```rust
+let query = "select (
+    insert Account {
+    username := <str>$0
+    }) {
+    id, 
+    username
+    };";
+let cannot_make_into_queryable_account: Result<QueryableAccount, _> =
+    client.query_required_single(query, &("SomeUserName",)).await;
+assert!(
+    format!("{cannot_make_into_queryable_account:?}")
+    .contains(r#"error: Some(WrongField { unexpected: "id", expected: "username" })"#);
```

Comment:
Yes, Display output from .unwrap_err() just shows `DescriptorMismatch`.

Re: the flow - this is one of the first hiccups I had with the macro and having this nota bene up front would have been helpful. (Want to avoid any new users concluding that the macro doesn't work and moving on to less slick methods to deserialize when all they have to do is match the query shape to the Rust struct)